### ISSUE TITLE
Exit 1 when upgrading to Ubuntu 21.04

### ISSUE THREAD

**Issue by Mystic8b** [microsoft/WSL#6942]
### Windows Build Number

Microsoft Windows [Version 10.0.19043.985]

### WSL Version

- [X] WSL 2
- [ ] WSL 1

### Kernel Version

Linux version 5.4.72-microsoft-standard-WSL2 (oe-user@oe-host) (gcc version 8.2.0 (GCC)) #1 SMP Wed Oct 28 23:40:43 UTC 2020

### Distro Version

Release:        20.10

### Other Software

VSCode 1.56.2 with Remote-WSL v0.56.2

### Repro Steps

Change Prompt to normal in `/etc/update-manager/release-upgrades`
`sudo apt dist-upgrade -y`
`sudo do-release-upgrade`

### Expected Behavior

The update successful

### Actual Behavior

Update failed

### Diagnostic Logs

```
Reading cache

Checking package manager
Reading package lists... Done
Building dependency tree
Reading state information... Done
Hit http://archive.ubuntu.com/ubuntu groovy InRelease
Hit http://archive.ubuntu.com/ubuntu groovy-updates InRelease
Hit http://archive.ubuntu.com/ubuntu groovy-backports InRelease
Hit http://security.ubuntu.com/ubuntu groovy-security InRelease
Fetched 0 B in 0s (0 B/s)
Reading package lists... Done
Building dependency tree
Reading state information... Done

Restoring original system state

Aborting
Reading package lists... Done
Building dependency tree
Reading state information... Done
=== Command terminated with exit status 1 (Fri May 14 10:31:40 2021) ===
```

**Comment by therealkenc** [comment]
The `do-release-upgrade` pretty much assumes `systemd` exists, daemons can be stopped and restarted with `systemctl`, and `snaps` work (among other things). That's beside generic problems with PPAs that can cause behavior like the above (for example [this guy](https://askubuntu.com/questions/1071646/restoring-original-system-state-aborting)).

The upgrade to 20.10 took here for me with `systemd` running. Chat about `systemd` is over in #994. 

![image](https://user-images.githubusercontent.com/7049700/118325638-9911e100-b4b8-11eb-8916-5efd562f79b6.png)

Another way to go is to replace `focal` with `groovy` manually in `/etc/apt/sources.list` with `nano` or `sed`, and do an `apt update; apt upgrade`. Rando guide for that [here](https://linuxize.com/post/how-to-upgrade-debian-9-stretch-to-debian-10-buster/), which for Stretch to Buster, but applies to updating `sources.list` in general on Debian derived distros.

But probably a better play still is to `wsl.exe --import` a Groovy tarball and forgo upgrade entirely. Which is what I would do IRL.

**Comment by Mystic8b** [comment]
Wow. This means that if it is necessary to change the distribution (if it appears in the win store), I need to reconfigure the working environment every time. It is a pain.
Are there any movements in this direction? Will WSL become more like real linux? The same systemd, etc. will ever appear?

**Comment by BilboTav** [comment]
I had similar issue, took me some time to debug it, but here is what I found out and how I resolved it in the end:

Upgrade tool uses some snap package system which is currently not supported in WSL. I disabled its usage by commenting one line in upgrade script and then it went fine. I upgraded from 20.04 --> 20.10, but I presume it would be same for any other upgrades.

===============================

Run `sudo do-release-upgrade`

After it fails with abort - go to `/tmp`, there should be some subfolder something like `/tmp/ubuntu-release-upgrader-49jp2ru6/` for example - where instance of upgrade tool is copied. Inside you will found out file called `DistUpgradeQuirks.py`. 

Fine lines 128, 129 and comment them:

```
#        if cache['snapd'].is_installed:
#            self._checkStoreConnectivity()
```

Then inside that folder run this command to prompt upgrade again from this edited instance of upgrade tool:
```
sudo ./dist-upgrade.py --frontend=DistUpgradeViewText
```

It should run fine now.
A bit complicated, but once the upgrades are done, there are done.

**Comment by Mystic8b** [comment]
@BilboTav Crazy😂 Now I can’t risk it, but I’ll try it soon. Thank you!

**Comment by guillotp** [comment]
Indeed, a little bit complicated but it works. thank you.
But now I have new issues with the xserver....

**Comment by BilboTav** [comment]
@guillotp Link the issue with Xserver here, maybe I dealt with that too.

**Comment by guillotp** [comment]
I am using VcXsrv for windows,
It is weird, it works with the option "mutliple windows". But with the option "One large windows" (option I used generally), I have now the following problem.
![image](https://user-images.githubusercontent.com/18597386/118723486-7efb3a00-b7fb-11eb-9cbb-74ac691b6e86.png)

**Comment by BilboTav** [comment]
I tried Your example with both of these settings generated by VcXsrv app:

```
<?xml version="1.0" encoding="UTF-8"?>
<XLaunch WindowMode="MultiWindow" ClientMode="NoClient" LocalClient="False" Display="-1" LocalProgram="xcalc" RemoteProgram="xterm" RemotePassword="" PrivateKey="" RemoteHost="" RemoteUser="" XDMCPHost="" XDMCPBroadcast="False" XDMCPIndirect="False" Clipboard="True" ClipboardPrimary="True" ExtraParams="" Wgl="True" DisableAC="False" XDMCPTerminate="False"/>
```

```
<?xml version="1.0" encoding="UTF-8"?>
<XLaunch WindowMode="Windowed" ClientMode="NoClient" LocalClient="False" Display="-1" LocalProgram="xcalc" RemoteProgram="xterm" RemotePassword="" PrivateKey="" RemoteHost="" RemoteUser="" XDMCPHost="" XDMCPBroadcast="False" XDMCPIndirect="False" Clipboard="True" ClipboardPrimary="True" ExtraParams="" Wgl="True" DisableAC="False" XDMCPTerminate="False"/>
```

App `startxfce4` worked fine on both settings.
Sorry, don't know... don't have much experience with VcXsrv. Maybe start a new issue for that.

**Comment by guillotp** [comment]
No worry. Thank you.

**Comment by CatInRl** [comment]
> I had similar issue, took me some time to debug it, but here is what I found out and how I resolved it in the end:
> 
> Upgrade tool uses some snap package system which is currently not supported in WSL. I disabled its usage by commenting one line in upgrade script and then it went fine. I upgraded from 20.04 --> 20.10, but I presume it would be same for any other upgrades.
> 
> ===============================
> 
> Run `sudo do-release-upgrade`
> 
> After it fails with abort - go to `/tmp`, there should be some subfolder something like `/tmp/ubuntu-release-upgrader-49jp2ru6/` for example - where instance of upgrade tool is copied. Inside you will found out file called `DistUpgradeQuirks.py`.
> 
> Fine lines 128, 129 and comment them:
> 
> ```
> #        if cache['snapd'].is_installed:
> #            self._checkStoreConnectivity()
> ```
> 
> Then inside that folder run this command to prompt upgrade again from this edited instance of upgrade tool:
> 
> ```
> sudo ./dist-upgrade.py --frontend=DistUpgradeViewText
> ```
> 
> It should run fine now.
> A bit complicated, but once the upgrades are done, there are done.

I came across this issue too. As I didn't use snap package, I just run `sudo apt remove snapd` to fix this issue. @Mystic8b @BilboTav

**Comment by moisespr123** [comment]
Pretty much my 21.04 setup is broken. Can't update packages because of systemd:

```
Fetched 3,137 kB in 1s (3,040 kB/s)
Setting up systemd (248.2-1ubuntu1) ...
systemd-machine-id-setup: error while loading shared libraries: libsystemd-shared-247.so: cannot open shared object file: No such file or directory
dpkg: error processing package systemd (--configure):
 installed systemd package post-installation script subprocess returned error exit status 127
Errors were encountered while processing:
 systemd
E: Sub-process /usr/bin/dpkg returned an error code (1)`
```

**Comment by new2f7** [comment]
> I just run `sudo apt remove snapd` to fix this issue.

Worked for me when upgrading from 18.04 to 20.04. (The [original issue](https://github.com/microsoft/WSL/issues/6952) has been closed.)

**Comment by throttlemeister** [comment]
There's two easy solutions to the original problem:
1) start snapd manually and run upgrade
2) remove snapd from system and run upgrade

Both should work just fine and as normal.

**Comment by telorand** [comment]
I know this is a closed issue, but in case anyone comes across this thread, do not use `do-release-upgrade` to upgrade your WSL Ubuntu version. The recommended method involves modifying your `sources.list` and `sources.list.d` with `sed` or `nano` and upgrading through `sudo apt update' and 'sudo apt upgrade` followed by `sudo apt full-upgrade`.

It would look something like this:
```bash
sudo sed -i 's/bionic/hirsute/g' /etc/apt/sources.list
sudo sed -i 's/bionic/hirsute/g' /etc/apt/sources.list.d/*.list
sudo apt update
sudo apt upgrade
sudo apt full-upgrade
sudo apt autoremove
```

- Details about the core concepts can be [found here](https://linuxize.com/post/how-to-upgrade-debian-9-stretch-to-debian-10-buster/#upgrade-debian-9-stretch-to-debian-10-buster).
- Some of the discussion about why you shouldn't use `do-release-upgrade` can be [found here](https://github.com/microsoft/WSL/issues/7054#issuecomment-859875805).

**Comment by hyoretsu** [comment]
@telorand What if we can get snapd working? (A [fully working workaround](https://github.com/microsoft/WSL/issues/5126#issuecomment-920507442) has been found for that)

**Comment by Serendipity-luck** [comment]
> I had similar issue, took me some time to debug it, but here is what I found out and how I resolved it in the end:
> 
> Upgrade tool uses some snap package system which is currently not supported in WSL. I disabled its usage by commenting one line in upgrade script and then it went fine. I upgraded from 20.04 --> 20.10, but I presume it would be same for any other upgrades.
> 
> ===============================
> 
> Run `sudo do-release-upgrade`
> 
> After it fails with abort - go to `/tmp`, there should be some subfolder something like `/tmp/ubuntu-release-upgrader-49jp2ru6/` for example - where instance of upgrade tool is copied. Inside you will found out file called `DistUpgradeQuirks.py`.
> 
> Fine lines 128, 129 and comment them:
> 
> ```
> #        if cache['snapd'].is_installed:
> #            self._checkStoreConnectivity()
> ```
> 
> Then inside that folder run this command to prompt upgrade again from this edited instance of upgrade tool:
> 
> ```
> sudo ./dist-upgrade.py --frontend=DistUpgradeViewText
> ```
> 
> It should run fine now. A bit complicated, but once the upgrades are done, there are done.

Thanks, It works for me.

**Comment by iaamhg** [comment]
> sudo apt remove snapd

Thanks a lot as It works for my case, and how on earth did you find out it is related with snapd, I checked all the logs under /var/log/dist-upgrade/ and find nothing usefull🤣

**Comment by BilboTav** [comment]
> > sudo apt remove snapd
> 
> Thanks a lot as It works for my case, and how on earth did you find out it is related with snapd, I checked all the logs under /var/log/dist-upgrade/ and find nothing usefull🤣

I debugged the python scrips looking for a place, where it exits with error. Pain in d***, but glad it helped some people.

**Comment by amirse** [comment]
> > > sudo apt remove snapd
> > 
> > 
> > Thanks a lot as It works for my case, and how on earth did you find out it is related with snapd, I checked all the logs under /var/log/dist-upgrade/ and find nothing usefullrofl
> 
> I debugged the python scrips looking for a place, where it exits with error. Pain in d***, but glad it helped some people.

Helped me as well. Thanks a lot!

**Comment by michael-newsrx** [comment]
Under "groovy".

```bash
sudo apt remove snapd
sudo do-release-upgrade --allow-third-party
```

success.

**Comment by hyoretsu** [comment]
You don't need to remove Snap. It's better if you don't, so you don't lose things like Docker. You just need to [make it work](https://github.com/microsoft/WSL/issues/5126#issuecomment-1001240706) before updating.

**Comment by michael-newsrx** [comment]
> You don't need to remove Snap. It's better if you don't, so you don't lose things like Docker. You just need to [make it work](https://github.com/microsoft/WSL/issues/5126#issuecomment-1001240706) before updating.

As snapd doesn't work, why keep it installed when its existence causes issues?

Also, the referenced method seems excessively convoluted.

One can always immediately reinstall snapd after the upgrade is finished.

**Comment by hyoretsu** [comment]
> > You don't need to remove Snap. It's better if you don't, so you don't lose things like Docker. You just need to [make it work](https://github.com/microsoft/WSL/issues/5126#issuecomment-1001240706) before updating.
> 
> As snapd doesn't work, why keep it installed when its existence causes issues?
> 
> Also, the referenced method seems excessively convoluted.
> 
> One can always immediately reinstall snapd after the upgrade is finished.

"The referenced method" is literally just running a command, upgrading from 20.04->21.04, then running another and upgrading from 21.04->21.10.

As snapd doesn't work, why reinstall it after the upgrade is finished when its existence causes issues? Said command makes Snapd work for that bash session.

**Comment by mattcargile** [comment]
Just went from `20.04` to `21.10` on Windows v1909 using @BilboTav's method successfully. The one change was I had to run the below to restart the upgrade ( i.e. the `.py` was in another folder ). I also made sure to run `apt update && apt upgrade && apt autoremove` prior the upgrade.

<code>sudo .**DistroUpgrade**/dist-upgrade.py --frontend=DistUpgradeViewText</code>

I wonder what lurks underneath, so to speak, because I didn't use @telorand's method using `sed` and `full-upgrade`. Is it the `microcode` failures? I did see those failures in my output I think. Not sure of the impact that entails though.

**Comment by internationils** [comment]
I found a simpler method in a few threads that worked... just `sudo apt remove snapd` and then it ran just fine and got me from the initial 20.04 directly to impish.

**Comment by bviktor** [comment]
> Wow. This means that if it is necessary to change the distribution (if it appears in the win store), I need to reconfigure the working environment every time. It is a pain. Are there any movements in this direction? Will WSL become more like real linux? The same systemd, etc. will ever appear?
> 
> **upd** It turned out that the store is still 20.04, which means that we can normally update only once every 2 years. This is ridiculous.

I'm not sure what you're talking about, the way @therealkenc explained is _the_ way all Debian and Ubuntu releases were done, before update-notifier even existed. Which is just a wrapper around apt anyway. In the background, it's pretty much just

```
sed -i.orig 's/focal/impish/g' /etc/apt/sources.list
sed -i.orig 's/focal/impish/g' /etc/apt/sources.list.d/*
apt update
apt dist-upgrade
```

There's no magic involved here, do-release-upgrade would do the exact same thing. You don't have to "reconfigure everything all the time", it's a one-time upgrade. To me it asked 4 questions, all about config files I did modify under /etc. So I had to go through the "trouble" of pressing Return 4 times to keep my versions instead of using the package versions. And then:

```
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 21.10
Release:        21.10
Codename:       impish
```

That's about it. So you guys seem to overthink this issue a little, just because you need to use the _original_ process instead of a fancy wrapper (do-release-upgrade). There's not much of an "issue" here.

**Comment by Mo01** [comment]
> sudo apt remove snapd
Worked for me I wasted a couple of hours before I came across your comment!!
